coderd/rbac/regosql/compile.go

```diff
@@ -0,0 +1,226 @@
+// clause. If the query equates to "true" then the user should have access.
+func ConvertRegoAst(cfg ConvertConfig, partial *rego.PartialQueries) (sqltypes.BooleanNode, error) {
+	if len(partial.Queries) == 0 {
+		// Always deny if there are no queries. This means there is no possible
+		// way this user can access these resources.
+		return sqltypes.Bool(false), nil
+	}
+
+	for _, q := range partial.Queries {
+		// An empty query in rego means "true". If any query in the set is
+		// empty, then the user should have access.
+		if len(q) == 0 {
+			// Always allow
+			return sqltypes.Bool(true), nil
+		}
+	}
+
+	var queries []sqltypes.BooleanNode
+	var builder strings.Builder
+	for i, q := range partial.Queries {
+		converted, err := convertQuery(cfg, q)
+		if err != nil {
+			return nil, xerrors.Errorf("query %s: %w", q.String(), err)
+		}
+
+		if i != 0 {
+			builder.WriteString("\n")
+		}
+		builder.WriteString(q.String())
+		queries = append(queries, converted)
+	}
+
+	// All queries are OR'd together. This means that if any query is true,
+	// then the user should have access.
+	sqlClause := sqltypes.Or(sqltypes.RegoSource(builder.String()), queries...)
+	// Always wrap in parens to ensure the correct precedence when combining with other
+	// SQL clauses.
+	return sqltypes.BoolParenthesis(sqlClause), nil
+}
+
+func convertQuery(cfg ConvertConfig, q ast.Body) (sqltypes.BooleanNode, error) {
+	var expressions []sqltypes.BooleanNode
+	for _, e := range q {
+		exp, err := convertExpression(cfg, e)
+		if err != nil {
+			return nil, xerrors.Errorf("expression %s: %w", e.String(), err)
+		}
+
+		expressions = append(expressions, exp)
+	}
+
+	// All expressions in a single query are AND'd together. This means that
+	// all expressions must be true for the user to have access.
+	return sqltypes.And(sqltypes.RegoSource(q.String()), expressions...), nil
+}
+
+func convertExpression(cfg ConvertConfig, e *ast.Expr) (sqltypes.BooleanNode, error) {
+	if e.IsCall() {
+		n, err := convertCall(cfg, e.Terms.([]*ast.Term))
+		if err != nil {
+			return nil, xerrors.Errorf("call: %w", err)
+		}
+
+		boolN, ok := n.(sqltypes.BooleanNode)
+		if !ok {
+			return nil, xerrors.Errorf("call %q: not a boolean expression", e.String())
+		}
+		return boolN, nil
+	}
+
+	// If it's not a call, it is a single term
+	if term, ok := e.Terms.(*ast.Term); ok {
+		ty, err := convertTerm(cfg, term)
+		if err != nil {
+			return nil, xerrors.Errorf("convert term %s: %w", term.String(), err)
+		}
+
+		tyBool, ok := ty.(sqltypes.BooleanNode)
+		if !ok {
+			return nil, xerrors.Errorf("convert term %s is not a boolean: %w", term.String(), err)
+		}
+
+		return tyBool, nil
+	}
+
+	return nil, xerrors.Errorf("expression %s not supported", e.String())
+}
+
+// convertCall converts a function call to a SQL expression.
+func convertCall(cfg ConvertConfig, call ast.Call) (sqltypes.Node, error) {
+	// Operator is the first term
+	op := call[0]
```

Comment:
Yes. But I can put a defensive check in case it is not used correctly.

The length 3 (really 2 after the op) is checked in `convertTerms(_, _ 2)`. So I know there is 2 args in the slice.